src/batou/lib/python.py

```diff
@@ -226,3 +229,53 @@ def update(self):
     @property
     def namevar_for_breadcrumb(self):
         return '{}=={}'.format(self.package, self.version)
+
+
+class AppEnv(Component):
+    """XXX docme"""
+
+    namevar = 'python_version'
+    env_hash = None
+
+    def configure(self):
+        self.requirements = open('requirements.lock', 'r').read()
+        hash_content = (
+            self.requirements +
+            self.python_version +
+            open(__file__, 'r').read())
+        hash_content = hash_content.encode('utf-8')
+        self.env_hash = hashlib.new('sha256', hash_content).hexdigest()
+        self.env_dir = os.path.join('.appenv', self.env_hash)
+        self.env_ready = os.path.join('.appenv', self.env_hash, 'appenv.ready')
+
+    def verify(self):
+        assert os.path.exists(self.env_ready)
```

Comment:
While i like it spelled this way, I don't think it's a good idea. Asserts may be optimised away, see https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement: "The current code generator emits no code for an assert statement when optimization is requested at compile time." (e.g. -O)

Reply:
I, too, remember reading about this the Python manual, ages ago. But is this really a realistic concern (as opposed to an idealistic one)? (I can see the headlines already: "batou does not work properly on PyPy"... ;)

The alternative `def assert_(condition): if not condition: raise AssertionError()` just does not have the same nice ring to it... we could play with synonyms though; `expect()` maybe?

Reply:
Well, I'd rather check that we run in a Python without -O. Pytest does something specific smart and -O is really not useful at all, it _only_ removes assertions and that isn't something that's actually done. batou as an application can make the assertion (heh) that it doesn't run in an optimizing interpreter.